### PR TITLE
[AArch64] Add assembly/disassembly for BFMOP4{A,S} (non-widening) instructions

### PR DESCRIPTION
**Pull request by momchil-velikov**
The new instructions are described in
https://developer.arm.com/documentation/ddi0602/2024-09/SME-Instructions

### CODE REVIEW

File: llvm/lib/Target/AArch64/SMEInstrFormats.td [hunk 1]
```diff
@@ -5126,3 +5126,77 @@ class sme2_luti4_vector_vg4_strided<bits<2> sz, bits<2> op, string mnemonic>
...
+  let Inst{4} = S;
+  let Inst{3-2} = 0b00;
+  let Inst{1-0} = ZAda;
+
+  let Constraints = "$ZAda = $_ZAda";
+}
+
+multiclass sme2_bfmop4as_widening<bit S, string mnemonic> {
+  // Single vectors
+  def _ZZ_S : sme2_bf16_fp32_quarter_tile_outer_product<0, 0, S, mnemonic, ZPR16Mul2_Lo, ZPR16Mul2_Hi>;
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SMEInstrFormats.td:5155]: In the previous patch Spencer pointed out that we are missing the M. So it should be _MZZ_S, _M2ZZ_S, _MZ2Z_S and _M2Z2Z_S

File: llvm/test/MC/AArch64/SME2p2/bfmop4as-non-widening.s [hunk 2]
```diff
@@ -0,0 +1,178 @@
+// RUN: llvm-mc -triple=aarch64 -show-encoding -mattr=+sme2p2,+sme-b16b16 < %s \
+// RUN:        | FileCheck %s --check-prefixes=CHECK-ENCODING,CHECK-INST
+// RUN: not llvm-mc -triple=aarch64 -show-encoding < %s 2>&1 \
+// RUN:        | FileCheck %s --check-prefix=CHECK-ERROR
+// RUN: llvm-mc -triple=aarch64 -filetype=obj -mattr=+sme2p2,+sme-b16b16 < %s \
+// RUN:        | llvm-objdump -d --mattr=+sme2p2,+sme-b16b16 - | FileCheck %s --check-prefix=CHECK-INST
+// RUN: llvm-mc -triple=aarch64 -filetype=obj -mattr=+sme2p2,+sme-b16b16 < %s \
+// RUN:        | llvm-objdump -d --mattr=-sme2 - | FileCheck %s --check-prefix=CHECK-UNKNOWN
```
Comment (CarolineConcatto) [inline llvm/test/MC/AArch64/SME2p2/bfmop4as-non-widening.s:8]: Can we fix this line to be --mattr=-sme2p2 -

**Review by CarolineConcatto — COMMENTED** [review]
Maybe we should add a commit message in this patch too.
The new instructions are described in
https://developer.arm.com/documentation/ddi0602/2024-09/SME-Instructions

File: llvm/lib/Target/AArch64/SMEInstrFormats.td (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SMEInstrFormats.td]: Done (on the other multiclass, though).

File: llvm/test/MC/AArch64/SME2p2/bfmop4as-non-widening.s (hunk 2 above)
Comment (momchil-velikov) [reply llvm/test/MC/AArch64/SME2p2/bfmop4as-non-widening.s]: Done.

**Review by jthackray — APPROVED** [review]
LGTM

**Pull request merged**